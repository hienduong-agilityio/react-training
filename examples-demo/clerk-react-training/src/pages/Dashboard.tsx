import { UserButton } from "@clerk/clerk-react";
import React from "react";
import { Link } from "react-router-dom";

const Dashboard: React.FC = () => {
  return (
    <>
      <nav className="flex items-center justify-between bg-white shadow px-6 py-4">
        <Link className="text-xl font-semibold text-gray-800" to="#">
          Dashboard
        </Link>
        <UserButton />
      </nav>

      <main className="flex h-[75vh] w-full items-center justify-center">
        <h1 className="text-4xl font-bold text-gray-700">User Dashboard</h1>
      </main>
    </>
  );
};

export default Dashboard;
