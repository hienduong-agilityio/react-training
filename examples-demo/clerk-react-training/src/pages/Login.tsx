import { SignIn } from "@clerk/clerk-react";

export default function Login() {
  return (
    <div className="flex items-center justify-center w-full h-screen">
      <SignIn signUpUrl="/signup" forceRedirectUrl="/dashboard" />
    </div>
  );
}
