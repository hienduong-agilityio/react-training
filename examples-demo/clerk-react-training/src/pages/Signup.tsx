import { SignUp } from "@clerk/clerk-react";

export default function Signup() {
  return (
    <div className="flex items-center justify-center w-full h-screen">
      <SignUp signInUrl="/login" forceRedirectUrl={"/dashboard"} />
    </div>
  );
}
