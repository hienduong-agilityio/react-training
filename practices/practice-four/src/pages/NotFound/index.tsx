// Libraries
import { Link } from 'react-router-dom';
import { usePageSeo } from '@/hooks';

// Constants
import { ROUTE } from '@/constants';

/**
 * NotFoundPage
 *
 * @returns {JSX.Element} - The NotFound element
 */
const NotFoundPage = (): JSX.Element => {
  usePageSeo({
    title: 'Page Not Found',
    description: 'The page you are looking for does not exist.',
    keywords: '404, not found, error',
    ogTitle: 'Page Not Found',
    ogDescription: 'The page you are looking for does not exist.'
  });

  return (
    <div className='flex flex-col h-screen justify-center items-center bg-gray-300'>
      <p className='text-[120px] font-extrabold text-gray-700'>404</p>
      <p className='text-2xl font-medium text-gray-600 mb-6'>Not Found</p>
      <Link
        to={ROUTE.ROOT}
        className='px-4 py-2 font-medium text-white bg-primary-500 rounded-md hover:bg-primary-600 transition-all duration-200 ease-in-out'
      >
        Go Home
      </Link>
    </div>
  );
};

export default NotFoundPage;
