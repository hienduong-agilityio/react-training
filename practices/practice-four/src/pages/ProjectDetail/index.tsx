// Libraries
import { Navigate, useParams } from 'react-router-dom';

// Components
import { ProjectDetailSkeleton } from '@/components';
import { ProjectDetailInfo } from '@/pages/ProjectDetail/ProjectDetailInfo';

// Hooks
import { usePageSeo, useProject } from '@/hooks';

// Interfaces
import { IProjectItemProps } from '@/interfaces';

const ProjectDetail = (): JSX.Element => {
  const { id } = useParams<{ id: string }>();
  const project = useProject({ id: id ?? '' });
  const { data: projectDetail, isQueryProjectDetailPending, error: projectDetailsError } = project;

  // Extract SEO data if projectDetail is valid
  const seoData =
    !isQueryProjectDetailPending && projectDetail && !Array.isArray(projectDetail)
      ? {
          title: `${projectDetail.projectName}`,
          description: projectDetail.projectName || `Details about project ${projectDetail.projectName}`,
          ogTitle: `${projectDetail.projectName}`,
          ogDescription: projectDetail.projectName || ''
        }
      : null;

  // Always call usePageSeo
  usePageSeo(
    seoData || {
      title: 'Project Details',
      description: 'A powerful tool for managing and tracking projects efficiently.'
    }
  );

  // Handle loading
  if (isQueryProjectDetailPending) return <ProjectDetailSkeleton />;

  // Handle not found
  if (!projectDetail) return <Navigate to='/404' replace />;

  // Handle error
  if (projectDetailsError) {
    return <div>Error loading project: {projectDetailsError.message}</div>;
  }

  const { manager: { managerName = 'Unknown', managerImage = '' } = {} } = projectDetail as IProjectItemProps;

  return <ProjectDetailInfo managerName={managerName} managerAvatar={managerImage} {...projectDetail} />;
};

export default ProjectDetail;
