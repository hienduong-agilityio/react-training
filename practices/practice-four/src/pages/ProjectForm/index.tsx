'use client';

// Libraries
import { useFormState } from 'react-dom';
import { Navigate, useNavigate } from 'react-router-dom';

// Hooks
import { initialProjectFormState, useProjectForm } from '@/hooks';

// Components
import { FormLayout, ProjectFormFields } from '@/components';

const ProjectFormPage = ({ title = '' }: { title: string }) => {
  const navigate = useNavigate();

  const { formAction, initialFormValues, isQueryProjectDetailPending, projectDetailsError } = useProjectForm(title);

  /**
   * useFormState is a hook that allows you to manage the state of a form.
   */
  const [formState, formActionDispatcher] = useFormState(formAction, initialProjectFormState);

  if (projectDetailsError) return <Navigate to='/404' replace />;

  const handleCancelForm = () => {
    if (window.history.length > 1) {
      navigate(-1);
    }
  };
  return (
    <FormLayout title={title}>
      <ProjectFormFields
        initialFormValues={initialFormValues}
        formErrorsMessages={formState.errors}
        isQueryProjectDetailPending={isQueryProjectDetailPending}
        title={title}
        handleCancelForm={handleCancelForm}
        formAction={formActionDispatcher}
      />
    </FormLayout>
  );
};

export default ProjectFormPage;
