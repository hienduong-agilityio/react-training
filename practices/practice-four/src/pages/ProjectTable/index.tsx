import { Suspense, useDeferredValue, useOptimistic, startTransition, useActionState, useState } from 'react';
import { useNavigate, useLocation, useSearchParams, Outlet } from 'react-router-dom';
import { useQuery, keepPreviousData } from '@tanstack/react-query';
import { useScan, Options } from 'react-scan';

// Components
import { ProjectSearch, Button, Spinner, ProjectTableManager } from '@/components';

// Constants
import { ROUTE, ROWS_PER_PAGE_OPTIONS } from '@/constants';

// Hooks
import { usePageSeo, useProject, useSearchProject } from '@/hooks';

// Services
import { getProjects } from '@/services';

// Types
import type { IProjectsQueryResult } from '@/interfaces';

const ProjectPage = () => {
  const navigate = useNavigate();
  const location = useLocation();
  const [searchParams] = useSearchParams();

  usePageSeo({
    title: 'Project Management',
    description: '"A powerful tool for managing and tracking projects efficiently.',
    keywords: 'project, management, tool',
    ogTitle: 'Project Management',
    ogDescription: '"A powerful tool for managing and tracking projects efficiently.',
    ogImage: 'https://example.com/og-image.jpg',
    ogUrl: 'https://project-managerment-table.netlify.app/'
  });

  const isChildRoute = location.state?.from;

  // Get values from URL
  const currentPage = parseInt(searchParams.get('page') ?? '1');
  const rowsPerPage = parseInt(searchParams.get('rowsPerPage') ?? '10');
  const initialField = searchParams.get('searchField') ?? 'projectName';
  const initialKeyword = searchParams.get(initialField) ?? '';

  // Search state hook
  const { searchField, searchKeyword, updateSearchTerm, updateSearchField } = useSearchProject(
    initialField,
    initialKeyword
  );

  const searchAction = async (_prevState: { searchValue: string }, formData: FormData) => {
    const keyword = formData.get('search') as string;

    updateSearchTerm(keyword);

    return { searchValue: keyword };
  };

  const searchFormStateTuple = useActionState(searchAction, {
    searchValue: initialKeyword
  });

  const formActionDispatcher = searchFormStateTuple[1];

  const [optimisticState, updateOptimistic] = useOptimistic({ currentPage, rowsPerPage }, (prev, next) => ({
    ...(prev ?? {}),
    ...(next ?? {})
  }));

  const deferredKeyword = useDeferredValue(searchKeyword);

  const {
    data: projectData,
    isQueryProjectsPending,
    error
  } = useProject({
    page: optimisticState.currentPage,
    rowsPerPage: optimisticState.rowsPerPage,
    filter: { [searchField]: deferredKeyword }
  });

  const allProjectsQuery: IProjectsQueryResult = useQuery({
    queryKey: ['projects', searchField, deferredKeyword],
    queryFn: () => getProjects({ filter: { [searchField]: deferredKeyword } }),
    placeholderData: keepPreviousData
  });

  const totalPages = Math.ceil((allProjectsQuery.data?.length ?? 1) / optimisticState.rowsPerPage);

  useScan({
    enabled: import.meta.env.DEV,
    name: 'ProjectPage',
    logProps: true,
    includeComponents: ['ProjectSearch', 'ProjectTableManager', 'Button'],
    rowsPerPage: optimisticState.rowsPerPage
  } as Options & { rowsPerPage: number });

  // Pagination handlers
  const handleRowsPerPageChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    const newRowsPerPage = parseInt(event.target.value, 10);

    startTransition(() => {
      updateOptimistic({ rowsPerPage: newRowsPerPage, currentPage: 1 });
    });

    const newParams = new URLSearchParams(searchParams.toString());

    newParams.set('rowsPerPage', newRowsPerPage.toString());
    newParams.set('page', '1');

    navigate(`?${newParams.toString()}`);
  };

  const handlePageChange = (newPage: number) => {
    startTransition(() => {
      updateOptimistic({ currentPage: newPage });
    });

    const newParams = new URLSearchParams(searchParams.toString());
    newParams.set('page', newPage.toString());

    navigate(`?${newParams.toString()}`);
  };

  // Delete modal state
  const [isDeleteModalOpen, setIsDeleteModalOpen] = useState(false);
  const [selectedProjectId, setSelectedProjectId] = useState('');

  const handleOpenDeleteProjectModal = (id: string) => {
    startTransition(() => {
      setSelectedProjectId(id);
      setIsDeleteModalOpen(true);
    });
  };

  const handleDeleteModalClose = () => {
    startTransition(() => {
      setIsDeleteModalOpen(false);
    });
  };

  // Option builder
  const rowPerPageOption = ROWS_PER_PAGE_OPTIONS.map((option) => (
    <option key={option.value} value={option.value}>
      {option.label}
    </option>
  ));

  const handleNewProjectClick = () => {
    navigate(ROUTE.ADD_PROJECT, { state: { from: location.pathname + location.search } });
  };

  const isPending = isQueryProjectsPending || allProjectsQuery.isFetching;

  if (error ?? allProjectsQuery.error) return <div>Error loading projects</div>;

  return (
    <div className='h-full'>
      {!isChildRoute ? (
        <>
          <div className='flex justify-between pt-5 pb-7 px-3'>
            <ProjectSearch
              searchField={searchField}
              updateSearchField={updateSearchField}
              initialKeyword={initialKeyword}
              formAction={formActionDispatcher}
            />
            <Button onClick={handleNewProjectClick}>New Project</Button>
          </div>

          <Suspense fallback={<Spinner />}>
            <ProjectTableManager
              isPending={isPending}
              allProjectsQuery={allProjectsQuery}
              projectData={Array.isArray(projectData) ? projectData : []}
              rowsPerPage={optimisticState.rowsPerPage}
              totalPages={totalPages}
              currentPage={optimisticState.currentPage}
              selectedProjectId={selectedProjectId}
              isDeleteModalOpen={isDeleteModalOpen}
              rowPerPageOption={rowPerPageOption}
              handleRowsPerPageChange={handleRowsPerPageChange}
              handlePageChange={handlePageChange}
              handleOpenDeleteProjectModal={handleOpenDeleteProjectModal}
              handleDeleteModalClose={handleDeleteModalClose}
            />
          </Suspense>
        </>
      ) : (
        <Outlet />
      )}
    </div>
  );
};

export default ProjectPage;
