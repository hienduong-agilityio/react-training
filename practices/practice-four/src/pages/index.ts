// Index Page files
export { default as NotFoundPage } from './NotFound';
export { default as ProjectDetail } from './ProjectDetail';
export { default as ProjectFormPage } from './ProjectForm';
export { default as ProjectPage } from './ProjectTable';
