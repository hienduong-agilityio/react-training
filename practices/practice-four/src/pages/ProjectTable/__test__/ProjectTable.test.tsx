// Libraries
import { render, screen } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { useLocation } from 'react-router-dom';

// Pages
import { ProjectPage } from '@/pages';

jest.mock('react-router-dom', () => ({
  useNavigate: jest.fn(),
  useLocation: jest.fn().mockReturnValue({ pathname: '/projects', search: '', state: null }),
  useSearchParams: jest.fn().mockReturnValue([new URLSearchParams('')]),
  Outlet: () => <div>Mock Outlet</div>
}));

jest.mock('@/components/ProjectTableManager', () => jest.fn(() => <div>ProjectTableManager</div>));

const renderWithQueryClient = (ui: JSX.Element) => {
  const queryClient = new QueryClient();
  return render(<QueryClientProvider client={queryClient}>{ui}</QueryClientProvider>);
};

describe('ProjectPage', () => {
  let locationMock: jest.Mock;

  beforeEach(() => {
    locationMock = useLocation as jest.Mock;
  });

  it('should match snapshot', () => {
    const { asFragment } = renderWithQueryClient(<ProjectPage />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render ProjectTableManager component when not in child route', () => {
    renderWithQueryClient(<ProjectPage />);
    const tableManager = screen.getByText('ProjectTableManager');
    expect(tableManager).toBeInTheDocument();
  });

  it('should render Outlet component when in child route', () => {
    locationMock.mockReturnValue({
      pathname: '/projects',
      search: '',
      state: { from: '/child-route' }
    });

    renderWithQueryClient(<ProjectPage />);
    const outlet = screen.getByText('Mock Outlet');
    expect(outlet).toBeInTheDocument();
  });
});
