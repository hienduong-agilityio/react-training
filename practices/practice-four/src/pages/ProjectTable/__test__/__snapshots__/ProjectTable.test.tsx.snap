// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectPage should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="h-full"
  >
    <div
      class="flex justify-between pt-5 pb-7 px-3"
    >
      <form
        action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
        class="flex h-auto"
      >
        <div
          class="relative inline-block text-left"
        >
          <select
            class="inline-flex items-center py-0 h-[33px] text-sm font-medium text-center border-0 rounded-none rounded-s-lg bg-white hover:border-none focus:ring-4 focus:outline-none focus:ring-primary-100 text-gray-700 shadow"
            role="combobox"
          >
            <option
              value="projectName"
            >
              Project name
            </option>
            <option
              value="status"
            >
              Project status
            </option>
          </select>
        </div>
        <div
          class="relative"
        >
          <img
            alt="Search"
            class="pl-4 p-2 absolute"
          />
          <div
            class="flex flex-col"
          >
            <input
              class="border gap-2 block outline-none focus:ring-4 focus:ring-primary-100 rounded-md py-[6px] px-3 text-gray-900 disabled:text-light disabled:bg-white shadow h-full w-full pl-9 rounded-none rounded-e-lg"
              id="searchInput"
              name="search"
              placeholder="Search"
              type="search"
              value=""
            />
          </div>
        </div>
        <button
          class="flex gap-2 items-center focus:ring-4 focus:ring-primary-100 rounded-md cursor-pointer disabled:opacity-45 bg-white border-2 text-black px-3 py-1 text-sm text-black ml-2 focus:bg-gray-200 hover:bg-gray-200 focus:outline-none"
          type="submit"
        >
          Search
        </button>
      </form>
      <button
        class="flex gap-2 items-center focus:ring-4 focus:ring-primary-100 rounded-md cursor-pointer disabled:opacity-45 border-white text-primary-400 hover:border-2 hover:border-primary-100 px-4 py-2 text-base text-black"
      >
        New Project
      </button>
    </div>
    <div>
      ProjectTableManager
    </div>
  </div>
</DocumentFragment>
`;
